### PR TITLE
[JENKINS-69183] It is impossible to update disabled plugins

### PR DESCRIPTION
**Pull request by benebsiny**
See [JENKINS-69183](https://issues.jenkins.io/browse/JENKINS-69183).

### Case 1 (related to current issue):
1. Install the old version of `git` plugin.
2. Disable `git`.
3. Update `git`.

**Result**:
* `git` is updated.
* `git` remains disabled.

### Case 2:
(`git` has dependency `git-client`)
1. Install the old version of `git` plugin.
2. Disable `git` and `git-client`.
3. Update `git`.

**Result**:
* `git` is updated.
* `git` and `git-client` remains disabled.

### Case 3 (related to [cause issue](https://issues.jenkins.io/browse/JENKINS-34494)):
1. Install `git-client`.
2. Disabled `git-client`.
3. Install `git`.

**Result**:
* `git-client` is enabled.

### Case 4 (Optional dependency):
(`matrix-auth` has optional dependency `cloudbees-folder`)
1. Install `cloudbees-folder`.
2. Disable `cloudbees-folder`.
3. Install `matrix-auth`.

**Result**
* `cloudbees-folder` remains disabled.

### Case 5 (Implied dependency):
(`matrix-auth` has implied dependency `JavaMail API`)
1. Install `JavaMail API`.
2. Disable `JavaMail API`.
3. Install `matrix-auth`.

**Result**
* `JavaMail API`  is enabled.

### Proposed changelog entries

- Fix the update of disabled plugin

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [X] (If applicable) Jira issue is well described
- [X] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change) and are in the imperative mood. [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  - Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadoc, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")` if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content-Security-Policy directives (see [documentation on jenkins.io](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`:

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are accurate, human-readable, and in the imperative mood
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/7040"></a>

### CODE REVIEW

File: core/src/main/java/hudson/model/UpdateSite.java
Comment (benebsiny) [reply core/src/main/java/hudson/model/UpdateSite.java:1588]: @daniel-beck  Actually not all callers pass `false`. [Line 1599](https://github.com/jenkinsci/jenkins/pull/7040/files/0e48fb6466746bb65afb35ca8bd4102c8dffc30a#diff-aeb05543bf9aa3e347768cef830d36f45cd049454384552ee02320cea0f3791bR1599) passes the current status of this plugin (dependent plugin), and it may be `true`.

**Review by basil — APPROVED** [review]

**Review by NotMyFault — APPROVED** [review]

**Comment by daniel-beck** [comment]
How does this behave with implied and optional dependencies? E.g. not `git` and `git-client`, but `matrix-auth` and `cloudbees-folder`, or `matrix-auth` and `sshd`?

**Comment by daniel-beck** [comment]
> What is the reason behind the asking of this question?

The plugin manager is a fragile mess.

> Is there some concern that this has not been fully tested?

Given this case isn't covered in the PR description, yes.

> If so, what is the testing you are requesting, specifically?

See PR description and replace the plugin names provided there with the plugin names provided in my comment.

> And what is the reason you haven't downloaded this PR and played around with it yourself to answer your own question?

I'd be happy to test but 2 approvals without any nonpositive feedback means this gets merged sooner than I can make the time to actually look at this in sufficient detail.

**Comment by basil** [comment]
Thanks for clarifying @daniel-beck. @benebsiny Can you please perform the testing @daniel-beck has requested and update the testing section of PR description?

**Comment by basil** [comment]
> So I am forced to comment my concerns without having first confirmed them.

Nobody is forcing you to say anything about a feeling you have that may or not be a real issue.

**Comment by daniel-beck** [comment]
> > So I am forced to comment my concerns without having first confirmed them.
> 
> Nobody is forcing you to say anything about a feeling you have that may or not be a real issue.

I've reported 31 regressions in core so far this year, so I'd rather comment than do that some more. And please don't suggest I do neither.

**Comment by basil** [comment]
> @benebsiny requested a review from @daniel-beck 11 days ago

Reiterating this request for @daniel-beck to confirm that the testing he requested was performed.

**Comment by basil** [comment]
@daniel-beck Do you intend to review this PR as requested?

**Review by daniel-beck — APPROVED** [review]
Test cases look good. Thanks for going through these cases and confirming behavior is as expected! I also tried the change locally and could not find a case that broke.

For a while I suspected problems, or at least inconsistent behavior, when a dependent plugin is newly installed and already disabled (vs. disabled and being updated), but that should be taken care of by https://github.com/jenkinsci/jenkins/pull/7042.

**Comment by timja** [comment]
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Review by basil — APPROVED** [review]

**Pull request merged**
